From 18ffcfd3d8dc44bf2b244d1b10660131cf407ae4 Mon Sep 17 00:00:00 2001
From: Shimin Zhou <shimingx.zhou@intel.com>
Date: Tue, 19 Jan 2016 10:12:28 +0800
Subject: [PATCH 423/429] spi: enable DMA with no trail mode

The Merrifield DMA is capable of emptying trailing bytes of
SPI FIFO, enable it to improve SPI transferring performance.

Signed-off-by: Shimin Zhou <shimingx.zhou@intel.com>
Signed-off-by: Zhenming Zhao <zhenmingx.zhao@intel.com>
---
 drivers/spi/intel_mid_ssp_spi.c | 10 ++++++++--
 1 file changed, 8 insertions(+), 2 deletions(-)

diff --git a/drivers/spi/intel_mid_ssp_spi.c b/drivers/spi/intel_mid_ssp_spi.c
index cbc993abd418..3199e612dc02 100644
--- a/drivers/spi/intel_mid_ssp_spi.c
+++ b/drivers/spi/intel_mid_ssp_spi.c
@@ -1111,6 +1111,11 @@ static int handle_message(struct ssp_drv_context *sspc)
 			} else
 				write_SSTO(timeout, reg);
 		}
+		else
+		{
+			write_SSTO(timeout, reg);
+		}
+
 		dev_dbg(dev, "transfer len:%d  n_bytes:%d  cr0:%x  cr1:%x",
 				sspc->len, sspc->n_bytes, cr0, cr1);
 
@@ -1385,7 +1390,7 @@ static int setup(struct spi_device *spi)
 		spin_lock_irqsave(&sspc->lock, flags);
 		sspc->cr1_sig = SSCR1_TSRE | SSCR1_RSRE;
 		sspc->mask_sr = SSSR_ROR | SSSR_TUR;
-		if (sspc->quirks & QUIRKS_DMA_USE_NO_TRAIL)
+/*		if (sspc->quirks & QUIRKS_DMA_USE_NO_TRAIL)	*/
 			sspc->cr1_sig |= SSCR1_TRAIL;
 	} else {
 		sspc->cr1_sig = SSCR1_TINTE;
@@ -1486,7 +1491,8 @@ static int intel_mid_ssp_spi_probe(struct pci_dev *pdev,
 			sspc->quirks |= QUIRKS_USE_PM_QOS |
 					QUIRKS_SRAM_ADDITIONAL_CPY;
 	}
-	sspc->quirks |= QUIRKS_DMA_USE_NO_TRAIL;
+	if (!(sspc->quirks & QUIRKS_PLATFORM_MRFL))
+		sspc->quirks |= QUIRKS_DMA_USE_NO_TRAIL;
 	if (ssp_cfg_is_spi_slave(ssp_cfg))
 		sspc->quirks |= QUIRKS_SPI_SLAVE_CLOCK_MODE;
 
-- 
2.37.3

